### PR TITLE
Use structures instead of a individual parameters for command callback

### PR DESCRIPTION
**Pull request by jspaith**
Move the callback for processing commands to taking a ton of arguments to moving them to structures.

So the bottom line here - the original callback function signature...
```c
    typedef int(*IOTHUB_CLIENT_COMMAND_CALLBACK_ASYNC)(
                    const char* componentName,
                    const char* commandName,
                    const unsigned char* payload,
                    size_t size,
                    const char* payloadContentType,
                    unsigned char** response,
                    size_t* responseSize,
                    void* userContextCallback);
```

... becomes new ...

```c
    typedef void(*IOTHUB_CLIENT_COMMAND_CALLBACK_ASYNC)(
                const IOTHUB_CLIENT_COMMAND_REQUEST* commandRequest,
                IOTHUB_CLIENT_COMMAND_RESPONSE* commandResponse,    
                void* userContextCallback);
```

... with the various fields, along with struct versioning, put into the new `IOTHUB_CLIENT_COMMAND_REQUEST` and `IOTHUB_CLIENT_COMMAND_RESPONSE`.

Why?
* If we ever add new fields for requests/responses, we can add them to a structure without doing `_Ex()` business on callbacks.  

* This brings us inline with embedded SDK and [FreeRTOS Middleware](https://github.com/Azure/azure-iot-middleware-freertos/blob/d8efbc0b987efee992e848f3f3a86a706ac3f3ad/source/include/azure_iot_hub_client.h#L156) design.

* This also brings us inline with the [C# proposal](https://github.com/Azure-Samples/azure-iot-samples-csharp/tree/preview/iot-hub/Samples/device/ConventionBasedOperations#using-convention-aware-api-new-2) of request|response objects.

* The 8 parameters the original has "feels" like too many.  We did this at 1st since it was following the existing device method pattern, but since we have custom code to process commands in core SDK it gives freedom to easily rearange calling pattern.

### CODE REVIEW

File: iothub_client/inc/iothub_client_core_common.h [hunk 1]
```diff
@@ -348,31 +348,55 @@ extern "C"
...
+
+    /** @brief    This struct specifies parameters of an incoming command response */  
+    typedef struct IOTHUB_CLIENT_COMMAND_RESPONSE_TAG
+    {
+        /** @brief   Version of the structure.  Currently must be IOTHUB_CLIENT_COMMAND_RESPONSE_STRUCT_VERSION_1. */
+        int structVersion;
+        /** @brief Response generated by the callback that will be sent to IoT Hub.  This must be allocated using malloc().  
+                   The SDK takes ownership of the buffer on callback return and will free() it when the SDK is done with it. */
+        unsigned char* payload;
+        /** @brief Number of bytes application is returning in response.  This should *not* include the terminating null. */
```
Comment (ericwol-msft) [inline iothub_client/inc/iothub_client_core_common.h:383]:
```suggestion
        /** @brief Number of bytes application is returning in payload.  This should *not* include the terminating null. */
```

File: iothub_client/inc/iothub_client_core_common.h [hunk 2]
```diff
@@ -348,31 +348,55 @@ extern "C"
...
+        size_t payloadLength;
+        /** @brief Status code of the command to return to the service.  This maps to an HTTP style status code. */
```
Comment (ericwol-msft) [inline iothub_client/inc/iothub_client_core_common.h:385]:
```suggestion
        /** @brief Status code of the command to return to the IoT Hub service.  This maps to an HTTP style status code. */
```

**Review by ericwol-msft — APPROVED** [review]

File: iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c [hunk 3]
```diff
@@ -222,88 +222,81 @@ static bool BuildMaxMinCommandResponse(unsigned char** response, size_t* respons
 // SetEmptyCommandResponse sets the response to be an empty JSON.  IoT Hub needs
 // legal JSON, regardless of error status, so if command implementation did not set this do so here.
 //
-static void SetEmptyCommandResponse(unsigned char** response, size_t* responseSize, int* result)
+static void SetEmptyCommandResponse(IOTHUB_CLIENT_COMMAND_RESPONSE* commandResponse)
 {
-    if ((*response = calloc(1, g_JSONEmptySize)) == NULL)
+    if ((commandResponse->payload = calloc(1, g_JSONEmptySize)) == NULL)
     {
         LogError("Unable to allocate empty JSON response");
```
Comment (momuno) [inline iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c:229]: nit: c sdk style logging -- with or without '.' at end of sentence?  discussion in emb c sdk was to have the period.

File: iothub_client/inc/iothub_client_core_common.h [hunk 4]
```diff
@@ -348,31 +348,55 @@ extern "C"
...
+        const char* componentName;
+        /** @brief   Name of the command associated with this request. */
+        const char* commandName;
+        /** @brief    Raw payload of the request.  This is NOT guaranteed to be a \0 terminated string. */
+        const unsigned char* payload;
+        /** @brief    Number of bytes of @p payload. */
+        size_t payloadLength;
+        /** @brief    Payload type of @p payload, if and only if it was explicitly indicated by IoT Hub.  This currently 
+                      will ALWAYS be NULL.  Future versions of the SDK may set this value based on metadata received from IoT Hub. */
+        const char* payloadContentType;
```
Comment (danewalton) [inline iothub_client/inc/iothub_client_core_common.h:369]: Nit: the idea of having this structure approach is that we can add values in the future without breaking people. Should we really have this if it isn't even supported right now (and might not)?

File: iothub_client/inc/iothub_client_core_common.h [hunk 5]
```diff
@@ -348,31 +348,55 @@ extern "C"
...
+    } IOTHUB_CLIENT_COMMAND_REQUEST;
+
+    /** @brief Current version of @p IOTHUB_CLIENT_COMMAND_RESPONSE structure.  */
+    #define IOTHUB_CLIENT_COMMAND_RESPONSE_STRUCT_VERSION_1 1
+
+    /** @brief    This struct specifies parameters of an incoming command response */
```
Comment (danewalton) [inline iothub_client/inc/iothub_client_core_common.h:375]: would this be outgoing?
```suggestion
    /** @brief    This struct specifies parameters of an outgoing command response */  
```

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c [hunk 6]
```diff
@@ -164,61 +164,71 @@ static int MethodCallback(const char* method_name, const unsigned char* payload,
...
         LogError("Payload content type is non-NULL but should have been NULL");
-        responseCode = METHOD_RESPONSE_ERROR;
+        commandResponse->statusCode = METHOD_RESPONSE_ERROR;
     }
-    else if (size != strlen(expectedMethodPayload))
+    else if (commandRequest->payloadLength != strlen(expectedMethodPayload))
     {
-        LogError("payload size incorect - expected %zu but got %zu", strlen(expectedMethodPayload), size);
-        responseCode = METHOD_RESPONSE_ERROR;
+        LogError("payload size incorect - expected %zu but got %zu", strlen(expectedMethodPayload), commandRequest->payloadLength);
```
Comment (danewalton) [inline iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:200]:
```suggestion
        LogError("payload size incorrect - expected %zu but got %zu", strlen(expectedMethodPayload), commandRequest->payloadLength);
```

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c [hunk 7]
```diff
@@ -164,61 +164,71 @@ static int MethodCallback(const char* method_name, const unsigned char* payload,
...
+        commandResponse->statusCode = METHOD_RESPONSE_ERROR;
+    }
+    else if (commandRequest->payloadContentType != NULL)
     {
         LogError("Payload content type is non-NULL but should have been NULL");
-        responseCode = METHOD_RESPONSE_ERROR;
+        commandResponse->statusCode = METHOD_RESPONSE_ERROR;
     }
-    else if (size != strlen(expectedMethodPayload))
+    else if (commandRequest->payloadLength != strlen(expectedMethodPayload))
```
Comment (danewalton) [inline iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:198]: nit: should we change `expectedMethodPayload` to `expectedCommandPayload`

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c [hunk 8]
```diff
@@ -164,61 +164,71 @@ static int MethodCallback(const char* method_name, const unsigned char* payload,
...
+    {
+        LogError("Command request structVersion wrong - expected %d but got %d", IOTHUB_CLIENT_COMMAND_REQUEST_STRUCT_VERSION_1, commandRequest->structVersion);
+        commandResponse->statusCode = METHOD_RESPONSE_ERROR;
     }
-    else if (payloadContentType != NULL)
+    else if (IOTHUB_CLIENT_COMMAND_REQUEST_STRUCT_VERSION_1 != commandResponse->structVersion)
+    {
+        LogError("Command response structVersion wrong - expected %d but got %d", IOTHUB_CLIENT_COMMAND_REQUEST_STRUCT_VERSION_1, commandResponse->structVersion);
+        commandResponse->statusCode = METHOD_RESPONSE_ERROR;
+    }
```
Comment (danewalton) [inline iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:192]: are these duplicates?

**Review by danewalton — APPROVED** [review]
I trust your judgement on the comment about the non-GA-maybe-in-the-future struct field :)

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c (hunk 8 above)
Comment (danewalton) [reply iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:192]: Ahh wait the second one should be **IOTHUB_CLIENT_COMMAND_RESPONSE_STRUCT_VERSION_1 **

File: iothub_client/inc/iothub_client_core_common.h (hunk 4 above)
Comment (jspaith) [reply iothub_client/inc/iothub_client_core_common.h:369]: You're too modest - this isn't a nit but something I need to do.

File: iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c (hunk 3 above)
Comment (jspaith) [reply iothub_client/samples/pnp/pnp_simple_thermostat/pnp_simple_thermostat.c:229]: This repo isn't consistent (of course :().  

But a quick PowerShell (and I think Regex is correct, eyeballing it the # seems in ballpark) shows about a 9:1 ratio of no "." versus yes "."  So following the more common suggestion suggests no '.".

```script
C:\iot2\azure-iot-sdk-c\iothub_client> $a = gci -Recurse |Select-String "LogError"
C:\iot2\azure-iot-sdk-c\iothub_client> $a.count
2534
C:\iot2\azure-iot-sdk-c\iothub_client> ($a |Select-String '\."').count
245
```

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c (hunk 8 above)
Comment (jspaith) [reply iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:192]: Ugh.  Fixing.

File: iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c (hunk 7 above)
Comment (jspaith) [reply iothub_client/tests/common_device_method_e2e/iothubclient_common_device_method_e2e.c:198]: I complained about FreeRTOS doxgen confusing this last week I recall so what goes around comes around....  Fixing.

**Review by ewertons — APPROVED** [review]

**Pull request merged**